engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/InstrumentorBuiltin.java

```diff
@@ -20,10 +20,39 @@
 public class InstrumentorBuiltin extends Node {
   @Child private ArrayLikeAtNode atNode = ArrayLikeAtNode.create();
 
+  private static ThreadLocal<java.util.function.Function<String, Object>> CACHE =
+      new ThreadLocal<>();
+
+  @CompilerDirectives.TruffleBoundary
+  public static void runWithCache(
+      java.util.function.Function<String, Object> cache, Runnable action) {
+    var prev = CACHE.get();
+    try {
+      CACHE.set(cache);
+      action.run();
+    } finally {
+      CACHE.set(prev);
+    }
+  }
+
+  @CompilerDirectives.TruffleBoundary
+  private static Object findUuid(Object uuid) {
+    var cache = CACHE.get();
+    return cache == null ? null : cache.apply(uuid.toString());
```

Comment:
Could we stop using `String` as the key and instead use plain Object (UUID?)?

Reply:
What do you mean by _plain object_? What kind of change you'd like to see? Please keep in mind, this is a builtin - we have only _interpreter objects_, we don't have any `java.util.UUID` instance.